pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,450 @@
+                  isValidMergeWindow = false;
+                  break;
+                }
+              }
+
+              if (!isValidMergeWindow) {
+                break;
+              }
+            }
+          }
+          // windowStartMs > preSelectedSegment.getEndTimeMs()
+          // Haven't find the first overlapping segment, continue to the next segment
+        }
+
+        if (windowStartMs != waterMarkMs) {
+          updateWatermarkMs(offlineTableName, mergeLevel, windowStartMs);
+        }
+        lowerMergeLevel = mergeLevel;
+        lowerMergeLevelWatermarkMs = windowStartMs;
+
+        if (!isValidMergeWindow) {
+          LOGGER.warn(
+              "Window with start: {} and end: {} of mergeLevel: {} is not a valid merge window, Skipping task generation: {}",
+              windowStartMs, windowEndMs, mergeLevel, taskType);
+          continue;
+        }
+
+        if (!hasUnMergedSegments && !selectedSegments.isEmpty()) {
+          // If current merge window overlaps with all the newest pre-selected segments, and these segments are all merged,
+          // we don't have a chance to bump up the watermark in previous for loop, handle the case here.
+          LOGGER.debug("No unmerged segments found for window [{}, {}), Update window to: [{}. {})", windowStartMs,
+              windowEndMs, windowStartMs + bucketMs, windowEndMs + bucketMs);
+          windowStartMs += bucketMs;
+          updateWatermarkMs(offlineTableName, mergeLevel, windowStartMs);
+        }
+
+        if (!hasUnMergedSegments) {
+          LOGGER.info("No unmerged segments found for mergeLevel:{} for table: {}, Skipping task generation: {}",
+              mergeLevel, offlineTableName, taskType);
+          continue;
+        }
+
+        // Create task configs
+        SegmentPartitionConfig segmentPartitionConfig = tableConfig.getIndexingConfig().getSegmentPartitionConfig();
+        if (segmentPartitionConfig == null) {
+          pinotTaskConfigs.addAll(createPinotTaskConfigs(selectedSegments, offlineTableName, taskConfigs, mergeConfig));
+        } else {
+          // For partitioned table, schedule separate tasks for each partition
+          Map<String, ColumnPartitionConfig> columnPartitionMap = segmentPartitionConfig.getColumnPartitionMap();
+          Preconditions.checkState(columnPartitionMap.size() == 1, "Cannot partition on multiple columns for table: %s",
+              tableConfig.getTableName());
+          Map.Entry<String, ColumnPartitionConfig> partitionEntry = columnPartitionMap.entrySet().iterator().next();
+          String partitionColumn = partitionEntry.getKey();
+
+          Map<Integer, List<OfflineSegmentZKMetadata>> partitionToSegments = new HashMap<>();
+          for (OfflineSegmentZKMetadata selectedSegment : selectedSegments) {
+            Set<Integer> partitions = selectedSegment.getPartitionMetadata().getPartitions(partitionColumn);
+            Preconditions.checkState(partitions.size() == 1,
```

Comment:
partitions size might not be 1. We can group segments without valid partition as one special group, and merge them together

Reply:
Added outlier list to handle this case.